torch/distributed/_tools/mem_tracker.py

```diff
@@ -0,0 +1,920 @@
+         values as the memory consumed in bytes.
+    """
+
+    def __init__(self, mod_fqn: str):
+        self.mod_fqn = mod_fqn
+        self.parameter_mem: int
+        self.buffer_mem: int
+        self.input_mem: int
+        self.output_mem: int
+        self.local_peak: Dict[torch.device, int] = {}
+        self.snapshots: Dict[_ModState, List[Dict[torch.device, Dict[str, int]]]] = {}
+
+
+class _WeakRefInfo:
+    """
+    Manages memory statistics and device attributes for tensor storages.
+    """
+
+    def __init__(
+        self, size: int, element_size: int, device: torch.device, reftype: _RefType
+    ) -> None:
+        """
+        Initializes the ``_WeakRefInfo`` object with tensor storage properties.
+
+        Args:
+        size (int): The number of elements in the tensor storage.
+        element_size (int): The size of each element in the tensor storage.
+        device (torch.device): The device on which the tensor is allocated.
+        reftype (_RefType): The reference type of the tensor.
+        """
+        self.size = size
+        self.element_size = element_size
+        self.reftype = reftype
+        self.device = device
+        self.mem_consumed = self._calculate_mem_consumed()
+
+    def _calculate_mem_consumed(self) -> int:
+        """
+        Calculates the memory consumed by the tensor storage, considering device-specific allocation rules.
+
+        Returns:
+        int: The memory consumed in bytes.
+        """
+        mem = self.size * self.element_size
+        if self.device.type == "cuda":
+            return math.ceil((mem) / _PYTORCH_MIN_ALLOCATE) * _PYTORCH_MIN_ALLOCATE
+        return mem
+
+    def update_mem_consumed(self, st: torch.UntypedStorage) -> int:
+        """
+        Updates and returns the memory consumed if the storage size has changed.
+
+        Args:
+        st (torch.UntypedStorage): The tensor storage to check for size updates.
+
+        Returns:
+        int: The updated memory consumed in bytes.
+        """
+        if st.size() != self.size:
+            self.size = st.size()
+            self.mem_consumed = self._calculate_mem_consumed()
+        return self.mem_consumed
+
+    @staticmethod
+    def get_untyped_storages(t: torch.Tensor) -> Set[torch.UntypedStorage]:
+        """
+        Recursively extracts untyped storages from a tensor or its subclasses.
+
+        Args:
+        t (torch.Tensor): The tensor to extract storages from.
+
+        Returns:
+        Set[torch.UntypedStorage]: A set of untyped storages.
+        """
+        unflattend_tensors = [t]
+        flattened_tensor_storages = set()
+        while len(unflattend_tensors) > 0:
+            obj = unflattend_tensors.pop()
+            if is_traceable_wrapper_subclass(obj):
+                attrs, _ = obj.__tensor_flatten__()  # type: ignore[attr-defined]
+                unflattend_tensors.extend([getattr(obj, attr) for attr in attrs])
+            else:
```

Comment:
nit: This if/else assumes a tensor is either a traceable wrapper subclass or a normal tensor that has a storage, so it would not work for a non-traceable wrapper subclass, a non-wrapper subclass, etc. I wonder if we should try to make this more robust somehow.

Reply:
If you have any suggestions wrt to this let me know. This is mainly due to my knowledge limitation.

Reply:
I will just add a warning that no storage found for the tensor subclass, tracker will not track its storage and file an issue if someone has a use-case for this.

Reply:
> I will just add a warning that no storage found for the tensor subclass, tracker will not track its storage and file an issue if someone has a use-case for this.

Sounds great to me!